ClientRuntimes/CSharp/ClientRuntime.Azure/Authentication/ActiveDirectoryApplicationTokenProvider.cs

```diff
@@ -0,0 +1,153 @@
+using System.Globalization;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Azure.Properties;
+using Microsoft.IdentityModel.Clients.ActiveDirectory;
+using Microsoft.Rest;
+
+namespace Microsoft.Azure.Authentication
+{
+    /// <summary>
+    /// Provides tokens for Azure Active Directory applications. 
+    /// </summary>
+    internal class ActiveDirectoryApplicationTokenProvider : ITokenProvider
+    {
+        private AuthenticationContext _authenticationContext;
+        private string _tokenAudience;
+        private string _clientId;
+        private ClientCredential _credential;
+        private string _type;
+
+        /// <summary>
+        /// Initializes a token provider for application credentials.
+        /// </summary>
+        /// <param name="domain">The domain or tenant id for the application.</param>
+        /// <param name="clientId">The client Id of the application in Active Directory.</param>
+        /// <param name="secret">The application secret, used for authentication.</param>
+        /// <param name="environment">The Azure environment to manage resources in.</param>
+        public ActiveDirectoryApplicationTokenProvider(string domain, string clientId, string secret, AzureEnvironment environment) 
+            : this(domain, clientId, secret, environment, null)
+        {
+        }
+
+        /// <summary>
+        /// Initializes a token provider for application credentials.
+        /// </summary>
+        /// <param name="domain">The domain or tenant id for the application.</param>
+        /// <param name="clientId">The client Id of the application in Active Directory.</param>
+        /// <param name="secret">The application secret, used for authentication.</param>
+        /// <param name="environment">The Azure environment to manage resources in.</param>
+        /// <param name="cache">The TokenCache to use during authentication.</param>
+        internal ActiveDirectoryApplicationTokenProvider(string domain, string clientId, string secret, AzureEnvironment environment, TokenCache cache)
+        {
+             if (string.IsNullOrWhiteSpace(secret))
+            {
+                throw new ArgumentOutOfRangeException("secret");
+            }
+
+            InitializeAuthenticationContext(domain, clientId, environment, cache);
+            this._credential = new ClientCredential(clientId, secret);
+            ValidateAuthenticationResult(Authenticate().Result);
+       }
+
+        /// <summary>
+        /// Returns the token type of the returned token.
+        /// </summary>
+        public string TokenType
+        {
+            get { return _type; }
+        }
+
+        /// <summary>
+        /// Gets an access token from the token cache or from AD authentication endpoint. 
+        /// Attempts to refresh the access token if it has expired.
+        /// </summary>
+        public virtual async Task<string> GetAccessTokenAsync(CancellationToken cancellationToken)
```

Comment:
I'd still like to be explicit that this is the access token, which is a string, and not the entire authentication token, which contains the refresh token and other metadata